### ISSUE TITLE
Non-overridable media

### ISSUE THREAD

**Issue by appgurueu** [minetest/minetest#11270]
## Problem

Media can be overridden by texture packs or sound packs. The former is frequently used to gain a PvP advantage by highlighting certain aspects or X-ray texture packs. A first approach introduced use_texture_alpha to counter this. This is however insufficient: partially transparent nodes like water or leaves are still affected by X-ray texture packs. For fair PvP, it is important that all players use (mostly) the same textures.

## Solutions

Allow marking media files as non-overridable.

## Alternatives

There are workarounds like the use of texture modifiers or the like, but they are pretty incomplete. A [better workaround](https://github.com/MT-CTF/capturetheflag/pull/882) "implements" this using `dynamic_add_media`, but should only be used for small textures for obvious reasons.

**Comment by sfan5** [comment]
How does Minecraft solve this?

**Comment by Emojigit** [comment]
Hacked clients can still change the texture of non-overridable media or nodes after this change.

**Comment by appgurueu** [comment]
> Hacked clients can still change the texture of non-overridable media or nodes after this change.

I don't quite see the point. Hacked clients can pretty much arbitrarily display visuals to their user, and even node data they receive: all mapblocks, entities, HUDs, media...

But that doesn't mean *the official client* should allow for easy *X-ray texture packs*.

**Comment by hlqkj** [comment]
Texture overrides also allow to mark nodes that are supposed to be non-distinguishable from surrounding ones, eg. damaging falling nodes which are on some survival servers.

However, just one thought. The point comes to be always the same: preventing overrides to make things more fair will only affect 'fair players' who uses the official client, and actually ends up penalising them against the ones who knows how to go with hacked ones.

For this to be really fair, that hated feature about 'allow server owners the choice to ban unofficial clients' should be inplemented too!

Otherwise, as said, this would imho render things even more unfair, considering how difficult is for the regular player to get/recompile an hacked client instead of overriding something...

**Comment by hlqkj** [comment]
Also, I forgot this: sorry for the bluntness, but I find quite hilarious reading that kind of comments from a coredev @sfan5 :)

**Comment by Jipok** [comment]
> How does Minecraft solve this?

Doesn't solve it in any way)

**Comment by rubenwardy** [comment]
> Also, I forgot this: sorry for the bluntness, but I find quite hilarious reading that kind of comments from a coredev @sfan5 :)

Clean room implementations require that we don't read Minecraft's source code, or have too much knowledge of its internals

**Comment by hlqkj** [comment]
> Clean room implementations require that we don't read Minecraft's source code, or have too much knowledge of its internals

If that wasn't sarcasm, then it's precisely what I mean: doing it like Minecraft could keep you from finding better solutions to implement imho (and maybe let Minecraft do ot like minetest!)

If that was sarcasm, then 'lol' and nevermind me :)

**Comment by SmallJoker** [comment]
>   or X-ray texture packs. A first approach introduced use_texture_alpha to counter this.

In the meantime. further `use_texture_alpha` modes were added. It's up to mods to make proper use of it. As for leaves and other transparent textures, we'd need yet another nodedef parameter to defined the minimal average transparency of the node to sanitize texture pack textures manually. Is it worth the effort?

**Comment by appgurueu** [comment]
I agree that given `use_texture_alpha` and texture modifiers this is somewhat low priority (technically you could just stuff textures in `[png` for example), and for simple cases (such as forcing a certain opacity) there are simple ways to do this, but I don't consider it fully resolved yet.

PvP games shouldn't be forced to allow the user to arbitrarily swap out textures, or to use dirty and possibly inefficient hacks to combat this.

**Comment by kromka-chleba** [comment]
An alternative solution would be sending position of a player only to clients with the player in view (not hidden behind nodes). This doesn't solve the other X-ray related problems.
I believe this feature could be quite expensive (ray casts and such) so it'd have to be opt-in.

**Comment by hlqkj** [comment]
I'd just like to point out that the issue is not limited to X-ray view, or being able to spot players behind solid nodes. The use case I mentioned earlier, for example, is about nodes that *need* to be different because of their mechanics, but also *must* look identical to others for those mechanics to work. Specifically, this case is about stable vs. unstable stone nodes (both placed by the mapgen) where the unstable version causes damage when falling on a player. As of now, the falling nodes can easily be avoided by exploiting the `override.txt` file in a texture pack.

Being able to customize games is a great feature of Minetest which in my opinion shall be preserved: it would just be handy for a PvP or PvE game developer to have ways to limit cheating, either by being able to mark some specific media as non-overridable, or to otherwise limit how such media can be overridden.

**Comment by sfan5** [comment]
As discussed at an informal meeting at FOSDEM this isn't low priority.
The API could be as simple as `core.mark_media_non_overrideable("mymod_foo.png")` (ideally with a better name) and the rest if just transferring this bit and handling it on the client.

**Comment by sfan5** [comment]
I just realized it's not obvious how this should interact with a texture packs `override.txt`.
It would make sense that when a non-overrideable image is included a texture string, this causes the texture string to not be able to overridden entirely.
This isn't easy to do properly as it is, but doing a substring match would work as a hack.
